lib/bolt/pal/yaml_plan/step/resources.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module Bolt
+  class PAL
+    class YamlPlan
+      class Step
+        class Resources < Step
+          def self.allowed_keys
+            super + Set['resources']
+          end
+
+          def self.required_keys
+            Set['target']
+          end
+
+          def initialize(step_body)
+            super
+            @resources = step_body['resources']
+            @normalized_resources = normalize_resources(@resources)
+          end
+
+          def self.validate(body, step_number)
+            super
+
+            body['resources'].each do |resource|
+              if resource['type'] || resource['title']
+                if !resource['type']
+                  err = "Resource declaration must include type key if title key is set"
+                  raise step_error(err, body['name'], step_number)
+                elsif !resource['title']
+                  err = "Resource declaration must include title key if type key is set"
+                  raise step_error(err, body['name'], step_number)
+                end
+              else
+                type_keys = (resource.keys - ['parameters'])
+                if type_keys.empty?
+                  err = "Resource declaration is missing a type"
+                  raise step_error(err, body['name'], step_number)
+                elsif type_keys.length > 1
+                  err = "Resource declaration has ambiguous type: could be #{type_keys.join(' or ')}"
+                  raise step_error(err, body['name'], step_number)
+                end
+              end
+            end
+          end
+
+          # What if this comes from a code block?
+          def normalize_resources(resources)
+            resources.map do |resource|
+              if resource['type'] && resource['title']
+                type = resource['type']
+                title = resource['title']
+              else
+                type, = (resource.keys - ['parameters'])
+                title = resource[type]
+              end
+
+              { 'type' => type, 'title' => title, 'parameters' => resource['parameters'] || {} }
+            end
+          end
+
+          def body
+            @body.merge('resources' => @normalized_resources)
+          end
+
+          def transpile
+            code = StringIO.new
+            code.print "  "
+            code.print "$#{@name} = " if @name
+
+            code.puts "apply(#{Bolt::Util.to_code(@target)}) {"
+
+            declarations = @normalized_resources.map do |resource|
+              type = resource['type']
```

Comment:
```suggestion
type = resource['type'].respond_to?(:value) ? resource['type'].value : resource['type']
```

Reply:
@nicklewis I just read your comment below. Maybe this is not what we actually want...

Reply:
With that change I get
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt plan convert Boltdir/site/test/plans/yaml_apply.yaml 
# WARNING: This is an autogenerated plan. It may not behave as expected.
plan test::yaml_apply(
  TargetSpec $servers,
  TargetSpec $lb,
  String $site_content = 'hello'
) {
  if get_targets($lb).size != 1 {
  fail("Must specify a single load balancer, not ${lb}")
}

  apply_prep([$servers, $lb])

  $server_apply = apply($servers) {
    class { "nginx_example::server":
      site_content => "${site_content}"
,
    }
  }
  $lb_targetspec = get_targets($lb)[0]

  $server_targetspec = get_targets($servers)[0]

  $lb_apply = apply($lb) {
    class { 'haproxy': }
    ->
    haproxy::listen { "nginx":
      collect_exported => 'false,',
      ipaddress => $lb_targetspec.facts['ipaddress'],
      ports => '80',
    }
    ->
    haproxy::balancermember { $server_targetspec.name:
      listening_service => 'nginx',
      server_names => $server_targetspec.host,
      ipaddresses => $server_targetspec.facts['ipaddress'],
      ports => '80',
      options => 'check',
    }
  }

  return $server_apply
}
```
which appears to run in to the discrepency between resolving all variables at plan scope vs compilation
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt plan run test::yaml_to_puppet_apply servers=p71tjeuwlftc81l.delivery.puppetlabs.net lb=cazdnnex9p1rwk1.delivery.puppetlabs.net site_content='wow apply from yaml' --no-host-key-check 
Starting: plan test::yaml_to_puppet_apply
Starting: plan test::yaml_to_puppet_apply
Starting: install puppet and gather facts on p71tjeuwlftc81l.delivery.puppetlabs.net, cazdnnex9p1rwk1.delivery.puppetlabs.net
Finished: install puppet and gather facts with 0 failures in 2.76 sec
Starting: apply catalog on p71tjeuwlftc81l.delivery.puppetlabs.net
Finished: apply catalog with 0 failures in 5.05 sec
Starting: apply catalog on cazdnnex9p1rwk1.delivery.puppetlabs.net
cazdnnex9p1rwk1.delivery.puppetlabs.net: Evaluation Error: Error while evaluating a Method call, 'facts' parameter 'target' expects a Target value, got String (file: /home/cas/working_dir/bolt/Boltdir/site/test/plans/yaml_to_puppet_apply.pp, line: 28, column: 34) on node cazdnnex9p1rwk1.delivery.puppetlabs.net
```

Reply:
I believe the direct cause of that is [BOLT-1157](https://tickets.puppetlabs.com/browse/BOLT-1157). But yes, the difference in evaluation time would happen to work around it in the YAML version in this case.

Reply:
ahh, yep.